File: npc_logic.py
"""NPC generation logic.

This file focuses on "how to build an NPC" and deliberately avoids the menu
loop or database setup. That separation is one of the key refactors that will
make a future GUI much easier to build.
"""

import random

from ai_tools import enhance_npc_with_ai
from constants import OCCUPATIONS, PERSONALITY_TRAITS
from storage import load_npc_names


def _get_race_display_name(race_key, race_data):
    """Turn an internal race key into something friendly for display."""
    if "display_name" in race_data:
        return race_data["display_name"]

    if "_" in race_key:
        parts = race_key.split("_")
        return f"{parts[0].capitalize()} ({parts[1].capitalize()})"

    return race_key.capitalize()


def generate_npc(use_ai=True, names_data=None):
    """Generate one NPC and return both text and structured data.

    Returning both forms is useful:
    - the formatted text is great for a CLI or text box
    - the dictionary is better for saving to a database or showing in a GUI
    """
    if names_data is None:
        names_data = load_npc_names()

    if not names_data:
        return "NPC name data is unavailable. Check npc_names.json.", None

    available_races = list(names_data.keys())
    race_key = random.choice(available_races)
    sex = random.choice(["male", "female"])

    first_name = random.choice(names_data[race_key][sex])
    surname = random.choice(names_data[race_key]["surnames"])
    full_name = f"{first_name} {surname}"
    race_display = _get_race_display_name(race_key, names_data[race_key])

    if use_ai:
        ai_details = enhance_npc_with_ai(full_name, race_display, sex)
        if ai_details:
            npc_description = f"""
NPC Generated (AI-Enhanced):
{'=' * 50}
Name: {full_name}
Race: {race_display}
Sex: {sex.capitalize()}

Appearance:
{ai_details.get('appearance', 'N/A')}

Personality:
{ai_details.get('personality', 'N/A')}

Occupation:
{ai_details.get('occupation', 'N/A')}

Motivation:
{ai_details.get('motivation', 'N/A')}

Secret:
{ai_details.get('secret', 'N/A')}

Signature Quote:
"{ai_details.get('quote', 'N/A')}"
"""
            npc_data = {
                "name": full_name,
                "race": race_display,
                "sex": sex.capitalize(),
                "appearance": ai_details.get("appearance", "N/A"),
                "personality": ai_details.get("personality", "N/A"),
                "occupation": ai_details.get("occupation", "N/A"),
                "motivation": ai_details.get("motivation", "N/A"),
                "secret": ai_details.get("secret", "N/A"),
                "quote": ai_details.get("quote", "N/A"),
            }
            return npc_description, npc_data

    job = random.choice(OCCUPATIONS)
    personality_trait = random.choice(PERSONALITY_TRAITS)
    npc_description = f"""
NPC Generated:
{'=' * 50}
Name: {full_name}
Race: {race_display}
Sex: {sex.capitalize()}
Occupation: {job}
Personality: {personality_trait}
"""
    return npc_description, None

File: foundry-module/pf2e-gm-toolkit/scripts/npc-logic.js
/**
 * NPC generation logic. Ported from npc_logic.py.
 * Returns structured data objects rather than formatted strings so the UI
 * can render them however it likes.
 */

/** Pick a random element from an array. */
function _npcRandom(arr) {
  return arr[Math.floor(Math.random() * arr.length)];
}

/**
 * Turn an internal race key into a display name.
 * Matches Python's _get_race_display_name().
 * @param {string} raceKey
 * @param {object} raceData  entry from npc_names.json
 */
GMTOOLKIT.getRaceDisplayName = function (raceKey, raceData) {
  if (raceData.display_name) return raceData.display_name;
  if (raceKey.includes("_")) {
    const [first, second] = raceKey.split("_");
    return `${first.charAt(0).toUpperCase() + first.slice(1)} (${second.charAt(0).toUpperCase() + second.slice(1)})`;
  }
  return raceKey.charAt(0).toUpperCase() + raceKey.slice(1);
};

/**
 * Generate a basic NPC (no AI). Returns a plain object with the NPC fields.
 * @param {object} namesData        contents of npc_names.json
 * @param {string[]|null} [occupationPool]  optional filtered list of occupations;
 *                                          pass null/undefined to use the full list
 */
GMTOOLKIT.generateBasicNPC = function (namesData, occupationPool) {
  const races = Object.keys(namesData);
  if (races.length === 0) return null;

  const raceKey = _npcRandom(races);
  const raceData = namesData[raceKey];
  const sex = _npcRandom(["male", "female"]);

  const firstName = _npcRandom(raceData[sex]);
  const surname = _npcRandom(raceData.surnames);
  const race = GMTOOLKIT.getRaceDisplayName(raceKey, raceData);

  /* Use the provided occupation pool, or fall back to the full list.
     Guard against an empty pool so generation never returns null. */
  const pool = (occupationPool && occupationPool.length > 0)
    ? occupationPool
    : GMTOOLKIT.OCCUPATIONS;

  return {
    name: `${firstName} ${surname}`,
    race,
    sex: sex.charAt(0).toUpperCase() + sex.slice(1),
    occupation: _npcRandom(pool),
    personality: _npcRandom(GMTOOLKIT.PERSONALITY_TRAITS),
    /* These fields are null for basic NPCs; AI fills them. */
    appearance: null,
    motivation: null,
    secret: null,
    quote: null,
    aiEnhanced: false,
  };
};

/**
 * Merge AI-generated details onto an existing basic NPC object.
 * @param {object} npc     object returned by generateBasicNPC
 * @param {object} aiData  object returned by GMTOOLKIT.enhanceNPCWithAI
 */
GMTOOLKIT.applyAIEnhancement = function (npc, aiData) {
  return {
    ...npc,
    appearance: aiData.appearance ?? null,
    personality: aiData.personality ?? npc.personality,
    occupation: aiData.occupation ?? npc.occupation,
    motivation: aiData.motivation ?? null,
    secret: aiData.secret ?? null,
    quote: aiData.quote ?? null,
    aiEnhanced: true,
  };
};
